ql/models/volatility/garch.cpp

```diff
@@ -71,7 +71,7 @@ namespace QuantLib {
             Real retval(0.0);
             Real sigma2 = 0;
             Real u2 = 0;
-            for (const auto r2 : r2_) {
+            for (const auto& r2 : r2_) {
```

Comment:
Is this because `Real` is expensive to copy in your library? If so, it's not great that by default users have to pay the price of a dereference (assuming that it's not compiled out) for a cheap to copy type (i.e. `double`) but I suppose it's not worth thinking too much about until it becomes a measurable overhead for someone.

Reply:
I don't think we should do this to improve this rare use case

Reply:
It's not expense to copy in our library, but this triggers a warning with Clang and we wanted to remove it. But point taken @pcaspers - we'll revert that and live with the warning, or go back to less strict warning levels as @ralfkonrad suggested.

Reply:
It turns out the `const auto` triggered the warning. As that does not have much effect inside this small scope (without `&`), we removed the `const`. This is also consistent with plenty of other range-based for loops throughout QuantLib.

Reply:
I'm just curious, what was the name of the clang warning that you saw, and can you explain why it is emitted for your library and not the main library? I've never heard of clang complaining about a superfluous const before so I'd like to understand a bit more about it.

Reply:
This is the [exact warning + context](https://github.com/xcelerit/qlxad/actions/runs/4581748312/jobs/8091498525#step:8:672):

```
QuantLib/ql/models/volatility/garch.cpp:102:29: warning: loop variable 'r2' creates a copy from
type 'xad::AReal<double> const' [-Wrange-loop-construct]
            for (const auto r2 : r2_) {
                            ^
QuantLib/ql/models/volatility/garch.cpp:102:18: note: use reference type 'xad::AReal<double> const &' to 
prevent copying
```

That's because in the [XAD build](https://github.com/xcelerit/qlxad/actions/runs/4581748312/jobs/8091498525#step:8:672) `auto` resolves to `xad::AReal<double>`, where the copy-constructor is no longer trivial. Taking a reference fixes it (as we had initially). As to why it only triggers with `const` and is not without it, we don't know. Guessing that Clang thinks it's a typo (forgotten `&`) with the `const` keyword, as `const` without `&` is fairly uncommon in this case.

Reply:
Taking a non-const copy allows you to do non-const operations on r in the body, which is a common use case and therefore clang does not emit a warning. Taking a const copy on the other hand does not make sense, since you can do exactly the same things with r as with a const reference. Therefore you can as well avoid the copying.